urlchecker/core/urlproc.py

```diff
@@ -146,13 +149,100 @@ def extract_urls(self):
         # collect all links from file (unique=True is set)
         self.urls = fileproc.collect_links_from_file(self.file_name)
 
-    def check_urls(self, urls=None, retry_count=1, timeout=5):
+    async def aysnc_url_check(self, url, retry_count, timeout, headers):
+        """
+        Asyncronous check function for one url.
+
+        Args:
+            - urls           (list) : a list of urls to check.
+            - retry_count    (int)  : a number of retries to issue (defaults to 1, no retry).
+            - timeout        (int)  : a timeout in minutes for blocking operations like the connection attempt.
+            - headers        (dict) : headers to use in the request.
+        """
+        # init do retrails and retrails counts
+        do_retry = True
+        rcount = retry_count
+
+        # we will double the time for retry each time
+        retry_seconds = 15
+
+        # With retry, increase timeout by a second
+        pause = timeout
+        saved_responses = []
+        saved_errors = []
+
+        try:
+            while rcount > 0 and do_retry:
+                response = None
+                async with aiohttp.ClientSession(headers=headers) as session:
+                    try:
```

Comment:
And there's another try here?

Reply:
according to aiohttp the errors can happen in the session or the get call. The reason why I went with this implementation is that it was the best one to check the urls correctly. Also the tests don't necessarily provide the best coverage of all the exception possibilities that's why I am skeptical about the alternatives. One particular error I kept getting and I am not sure how to react to it other than increasing the pause variable was `[429 too many requests]` and that is why I tried to diversify the urls in the test files (that way I won't overwhelm the server).

Reply:
Ah just found this comment, it was in the hidden/scrunched up section (so my control F didn't see it!)